dotCMS/src/main/java/com/dotmarketing/portlets/browser/ajax/BrowserAjax.java

```diff
@@ -740,6 +740,58 @@ public boolean copyFolder ( String inode, String newFolder ) throws Exception {
         return true;
     }
 
+    private Optional<String> moveFolderWhenDestinationDoesNotExists(final String newFolder,
+																	final Folder currentFolder,
+																	final User user,
+																	final boolean respectFrontendRoles,
+																	final String errorString) throws Exception {
+
+		final Host parentHost = this.hostAPI.find(newFolder, user, respectFrontendRoles);
+
+		if (!permissionAPI.doesUserHavePermission(currentFolder, PERMISSION_WRITE, user)
+				|| !permissionAPI.doesUserHavePermission(parentHost, PERMISSION_WRITE, user)) {
+
+			throw new DotRuntimeException( "The user doesn't have the required permissions." );
+		}
+
+		if (!this.folderAPI.move(currentFolder, parentHost, user, respectFrontendRoles)) {
+			//A folder with the same name already exists on the destination
+			return Optional.ofNullable(errorString);
+		}
+
+		this.addRefreshIndexCommitListener(null, parentHost, currentFolder);
+		return Optional.empty();
+	}
+
+	private Optional<String> moveFolderToExistingDestination(final String newFolder,
+															 final Folder currentFolder,
+															 final User user,
+															 final boolean respectFrontendRoles,
+															 final String errorString) throws Exception {
+
+		final Folder parentFolder = this.folderAPI.find(newFolder, user, false);
+
+		if (!permissionAPI.doesUserHavePermission( currentFolder, PERMISSION_WRITE, user )
+				|| !permissionAPI.doesUserHavePermission( parentFolder, PERMISSION_WRITE, user )) {
+
+			throw new DotRuntimeException( "The user doesn't have the required permissions.");
+		}
+
+		if (parentFolder.getInode().equalsIgnoreCase(currentFolder.getInode()) || //Trying to move a folder over itself
+				this.folderAPI.isChildFolder(parentFolder, currentFolder)) {    //Trying to move a folder over one of its children
+
+			return Optional.ofNullable(errorString);
+		}
+
+		if (!folderAPI.move(currentFolder, parentFolder, user, respectFrontendRoles)) { //A folder with the same name already exists on the destination
+
+			return Optional.ofNullable(errorString);
+		}
+
+		this.addRefreshIndexCommitListener(parentFolder,null, currentFolder );
+		APILocator.getPermissionAPI().resetPermissionReferences(currentFolder);
```

Comment:
It seems like all this should be under the folderAPI - including user permission check, cache flush and permission reset